File: 0-webhook-signing-and-retry/0-typescript/src/emitter.controller.ts
import { Body, Controller, Get, NotFoundException, Param, Post } from "@nestjs/common"
import { EmitterService } from "./emitter.service"

interface PublishDto {
    type: string
    payload: unknown
    targetUrl: string
}

@Controller("api/events")
export class EmitterController {
    public constructor(private readonly emitter: EmitterService) {}

    @Post("publish")
    public publish(@Body() body: PublishDto): unknown {
        return this.emitter.publish({ type: body.type, payload: body.payload, targetUrl: body.targetUrl })
    }

    @Get(":eventId")
    public get(@Param("eventId") eventId: string): unknown {
        const record = this.emitter.get(eventId)
        if (record === undefined) {
            throw new NotFoundException(`event ${eventId} not found`)
        }
        return record
    }
}

File: 0-webhook-signing-and-retry/1-java/src/main/java/com/starci/webhook/EmitterController.java
package com.starci.webhook;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

/** Emitter HTTP surface; only registered when ROLE != receiver. */
@RestController
@RequestMapping("/api/events")
@ConditionalOnProperty(name = "ROLE", havingValue = "emitter", matchIfMissing = true)
public class EmitterController {

    private final EmitterService emitter;

    public EmitterController(EmitterService emitter) {
        this.emitter = emitter;
    }

    @PostMapping("/publish")
    public ResponseEntity<Records.EventRecord> publish(@RequestBody Records.PublishDto body) throws Exception {
        return ResponseEntity.status(HttpStatus.CREATED).body(emitter.publish(body));
    }

    @GetMapping("/{eventId}")
    public ResponseEntity<?> get(@PathVariable String eventId) {
        Records.EventRecord record = emitter.get(eventId);
        if (record == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", "event " + eventId + " not found", "statusCode", 404));
        }
        return ResponseEntity.ok(record);
    }
}
